### PR TITLE
[LTC] Upstream utils in computation_client

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69815
* #69716
* __->__ #69621

Summary: Upstream the following utils
- metrics.h
- multi_wait.h
- thread_pool.h
- unique.h

Differential Revision: [D32957629](https://our.internmc.facebook.com/intern/diff/D32957629)

### CODE REVIEW

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32957629).

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32957629).

**Comment by desertfire** [comment]
@JackCaoG , is pytorch/xla already using code from LTC? I wonder how this PR can trigger a XLA test failure?

**Comment by JackCaoG** [comment]
@desertfire It is a pt/xla head issue, I just disabled the test and will look into it a bit more. Don't worry about it for this pr

File: torch/csrc/lazy/core/config.h [hunk 1]
```diff
@@ -1,4 +1,7 @@
 #pragma once
 #include <c10/util/Flags.h>
 
+C10_DECLARE_int(torch_lazy_io_thread_pool_size);
```
Comment (alanwaketan) [inline torch/csrc/lazy/core/config.h:4]: I will let @wconstab comment on these.

**Review by alanwaketan — APPROVED** [review]
Otherwise, LGTM.

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32957629).

File: torch/csrc/lazy/core/config.h (hunk 1 above)
Comment (desertfire) [reply torch/csrc/lazy/core/config.h:4]: Do you mean how to setup them through environment variables?

Comment (wconstab) [reply torch/csrc/lazy/core/config.h:4]: I'm not sure what @alanwaketan means.

I made a bad assumption that all the declared flags would automatically be available as env vars or cmd line options.  I think this is true for gflags in some scenarios, but, i couldn't figure out how to set the env vars in torch so far.

So ultimately, i think we'll have to
 - convert some of these to pybindings and decide they don't need env vars at all
 - for ones that need env vars, somehow add support - either switch back to getenv, or, figure out how to turn on env parsing for C10_DECLARE flags.
 
 It probably doesn't hurt to keep following this pattern of moving all the envs into one config file though.  Might make it easier later on to make the above changes all-at-once?

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32957629).

**Review by wconstab — APPROVED** [review]

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32957629).

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32957629).

**Pull request closed without merging**